src/main/scala/amba/axi4/Deinterleaver.scala

```diff
@@ -9,38 +9,61 @@ import freechips.rocketchip.config.Parameters
+  * be from the same burst transaction, until the burst is complete.
+  *
+  * @param maxReadBytes is the maximum supported read burst size that this adapter
+  *   has been provisioned to support.
+  * @param buffer is the internal buffering to provide in the case where no deinterleaving is required.
+  */
 class AXI4Deinterleaver(maxReadBytes: Int, buffer: BufferParams = BufferParams.default)(implicit p: Parameters) extends LazyModule
 {
-  require (maxReadBytes >= 1 && isPow2(maxReadBytes))
+  require (maxReadBytes >= 1, s"AXI4Deinterleaver: maxReadBytes must be at least 1, not $maxReadBytes")
+  require (isPow2(maxReadBytes), s"AXI4Deinterleaver: maxReadBytes must be a power of two, not $maxReadBytes")
+
+  private def beats(slave: AXI4SlavePortParameters): Int =
+    (maxReadBytes+slave.beatBytes-1) / slave.beatBytes
+
+  // Nothing to do if R channel only uses a single beat
+  private def nothingToDeinterleave(slave: AXI4SlavePortParameters): Boolean =
+    beats(slave) <= 1
 
   val node = AXI4AdapterNode(
     masterFn = { mp => mp },
     slaveFn  = { sp => sp.copy(slaves = sp.slaves.map(s => s.copy(
       supportsRead = s.supportsRead.intersect(TransferSizes(1, maxReadBytes)),
       interleavedId = Some(0))))
-  })
+  }) {
+    override def circuitIdentity = edges.out.map(_.slave).forall(nothingToDeinterleave)
+  }
 
   lazy val module = new LazyModuleImp(this) {
     (node.in zip node.out) foreach { case ((in, edgeIn), (out, edgeOut)) =>
       val endId = edgeOut.master.endId
-      val beatBytes = edgeOut.slave.beatBytes
-      val beats = (maxReadBytes+beatBytes-1) / beatBytes
+      val beats = beats(edgeOut.slave)
 
-      // This adapter leaves the control + write paths completely untouched
+      // This adapter passes through the AR/AW control + W/B write data channels
       out.ar :<> in.ar
       out.aw :<> in.aw
       out.w :<> in.w
       in.b :<> out.b
 
-      if (beats <= 1) {
-        // Nothing to do if only single-beat R
+      // Only the R channel has the possibility of being changed
+      if (nothingToDeinterleave(edgeOut.slave)) {
         in.r.asInstanceOf[ReadyValidIO[AXI4BundleR]] :<> buffer.irrevocable(out.r)
       } else {
+        // We only care to deinterleave ids that are actually in use
+        val maxFlightPerId = Seq.tabulate(endId) { i =>
+          edgeOut.master.masters.find(_.id.contains(i)).flatMap(_.maxFlight).getOrElse(0)
+        }
+
         // Queues to buffer R responses
-        val qs = Seq.tabulate(endId) { i =>
-          val depth = edgeOut.master.masters.find(_.id.contains(i)).flatMap(_.maxFlight).getOrElse(0)
-          if (depth > 0) {
-            val q = Module(new Queue(out.r.bits.cloneType, beats))
+        val qs = maxFlightPerId.map { mf =>
+          if (mf > 0) {
+            val q = Module(new Queue(out.r.bits.cloneType, depth = beats))
             q.suggestName(s"queue_${i}")
             q.io
           } else {
```

Comment:
This and following block looks like it should just use the `Queue` factory method which is just a Wire when the size is 0. Other than the assertion that is.

Reply:
Hmm. Do you think it is actually better to unconditionally wire the interface through while applying some assertions to it for some ids than to feed the `DontCare` tieoff into the downstream logic?

Reply:
I don't know, I think having the assertion is good. You could use the Queue factory method and still check the size to see about adding the assertion but YMMV.

Reply:
Yeah I understand how to keep having the assertion, my question is more about QoR in the downstream circuitry. Like
is injecting "tied off" `DontCare` constants actually better or worse than just connecting all those wires to the external IO (but asserting failing if that IO is used in certain ways). The former sounds like there will be more muxes but with one side being a constant... I'm not sure I have time to evaluate this at the moment so am inclined to leave it alone for the moment

Reply:
Yeah I think it's fine as is.